### PR TITLE
[SPARK-10947] [SQL] With schema inference from JSON into a Dataframe, add option to infer all primitive object types as strings

### PR DESCRIPTION
**Pull request by stephend-realitymine**
Currently, when a schema is inferred from a JSON file using sqlContext.read.json, the primitive object types are inferred as string, long, boolean, etc.

However, if the inferred type is too specific (JSON obviously does not enforce types itself), this can cause issues with merging dataframe schemas.

This pull request adds the option "primitivesAsString" to the JSON DataFrameReader which when true (defaults to false if not set) will infer all primitives as strings.

Below is an example usage of this new functionality.

```
val jsonDf = sqlContext.read.option("primitivesAsString", "true").json(sampleJsonFile)

scala> jsonDf.printSchema()
root
|-- bigInteger: string (nullable = true)
|-- boolean: string (nullable = true)
|-- double: string (nullable = true)
|-- integer: string (nullable = true)
|-- long: string (nullable = true)
|-- null: string (nullable = true)
|-- string: string (nullable = true)
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 1]
```diff
@@ -632,6 +632,39 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
+    val jsonDF = sqlContext.read.option("primitivesAsString", "true").json(path)
+
+    val expectedSchema = StructType(
+      StructField("bigInteger", DecimalType(20, 0), true) ::
+      StructField("boolean", BooleanType, true) ::
+      StructField("double", DoubleType, true) ::
+      StructField("integer", LongType, true) ::
+      StructField("long", LongType, true) ::
+      StructField("null", StringType, true) ::
+      StructField("string", StringType, true) :: Nil)
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Looks like we need to change all of these data types to `StringType`, right?

Also, can you add a test with complex types (`StructType` and `ArrayType`) to make sure we preserve the structure?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 2]
```diff
@@ -1262,4 +1299,4 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
       )
     }
   }
-}
+}
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Add a newline.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala [hunk 3]
```diff
@@ -103,11 +107,14 @@ private[sql] object InferSchema {
...
         var elementType: DataType = NullType
         while (nextUntil(parser, END_ARRAY)) {
-          elementType = compatibleType(elementType, inferField(parser))
+          elementType = compatibleType(elementType, inferField(parser, primitivesAsString))
         }
 
         ArrayType(elementType)
 
+      case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) if primitivesAsString => StringType
+      case (VALUE_TRUE | VALUE_FALSE) if primitivesAsString => StringType
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala]: Add a newline between these two cases?

**Comment by yhuai** [comment]
ok to test

**Comment by yhuai** [comment]
Thank you @stephend-realitymine for working on it! Overall, the change in infer schema looks good. I left a comment at the test part.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 1 above)
Comment (stephend-realitymine) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Yes you are correct this should be ```StringType``` I will also add the test for complex types as you have suggested.

**Comment by yhuai** [comment]
LGTM. Thank you for working on it. Merging to master.

**Comment by yhuai** [comment]
@stephend-realitymine Can you ping me in the jira (https://issues.apache.org/jira/browse/SPARK-10947)? So, we can assign it to you.

**Pull request closed without merging**